beep/protocol_tools/maccor_to_arbin.py

```diff
@@ -0,0 +1,407 @@
+                    blank_step['m_szExtCtrlValue1'] = step_abs['StepValue']
+                    blank_step['m_szExtCtrlValue2'] = "0"
+                elif step_abs['StepType'] == 'Dischrge':
+                    blank_step['m_szCtrlValue'] = '-' + step_abs['Limits']['Current']
+                    blank_step['m_szExtCtrlValue1'] = step_abs['StepValue']
+                    blank_step['m_szExtCtrlValue2'] = "0"
+            elif step_abs['Limits'] is None:
+                blank_step['m_szStepCtrlType'] = "Voltage(V)"
+                if step_abs['StepType'] == ' Charge ':
+                    blank_step['m_szCtrlValue'] = step_abs['StepValue']
+                elif step_abs['StepType'] == 'Dischrge':
+                    blank_step['m_szCtrlValue'] = step_abs['StepValue']
+            else:
+                raise ValueError("Unable to set m_szStepCtrlType for voltage")
+
+        # Rest control mode
+        elif step_abs['StepMode'] == '        ' and step_abs['StepType'] == '  Rest  ':
+            if step_abs['Limits'] is None:
+                blank_step['m_szStepCtrlType'] = "Rest"
+            else:
+                raise ValueError("Unable to set m_szStepCtrlType for voltage")
+
+        # Flow control steps
+        elif step_abs['StepMode'] == '        ' and step_abs['StepType'] in [' Loop 1 ', '  Do 1  ',
+                                                                             ' Loop 2 ', '  Do 2  ',
+                                                                             'AdvCycle', '  End   ']:
+            if step_abs['StepType'] == 'AdvCycle':
+                blank_step['m_szStepCtrlType'] = "Set Variable(s)"
+                blank_step['m_szCtrlValue1'] = '0'
+                blank_step['m_szExtCtrlValue1'] = '1'
+                blank_step['m_szExtCtrlValue1'] = '0'
+            elif 'Loop' in step_abs['StepType']:
+                loop_counter = int(re.search(r'\d+', step_abs['StepType']).group())
+                blank_step['m_szStepCtrlType'] = "Set Variable(s)"
+                blank_step['m_szCtrlValue1'] = '0'
+                blank_step['m_szExtCtrlValue1'] = str(2 ** loop_counter)
+                blank_step['m_szExtCtrlValue1'] = '0'
+                assert isinstance(step_abs['Ends']['EndEntry'], OrderedDict)
+                loop_addendum = OrderedDict([('EndType', 'Loop Addendum'), ('Oper', '< '),
+                                            ('Step', step_flow_ctrl[step_index].split('-')[0]),
+                                             ('Value', step_abs['Ends']['EndEntry']['Value'])])
+                step_abs['Ends']['EndEntry'] = [loop_addendum, step_abs['Ends']['EndEntry']]
+
+            elif 'Do' in step_abs['StepType']:
+                loop_counter = int(re.search(r'\d+', step_abs['StepType']).group())
+                blank_step['m_szStepCtrlType'] = "Set Variable(s)"
+                blank_step['m_szCtrlValue1'] = str(2 ** (loop_counter + 15))
```

Comment:
Is this hard-coded offset arbin config-specific, and not expected to change?

Reply:
Yeah, Arbin is using a 20-bit value for control and masking off specific portions for each of the different actions (reset, increment, decrement) and using different bits for each of the different counters